rfc/rfc-92/rfc-92.md

```diff
@@ -0,0 +1,102 @@
+
+## Background
+
+Expanding further, there are plenty of valid technical reasons on why Hudi should allow different storage layouts, under the upper layer reader/writer and table services implementations.
+
+1. We have use-cases, for cloud-native/high performance implementations of timeline (\`HoodieTimeline\`) and metadata (\`HoodieMetadata\` interface). In our use-case, we would like to explore backing them using NoSQL datastore like DynamoDB, for ultra-low latency queries.
+2. Hudi already supports [different](https://github.com/apache/hudi/blob/master/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/storage/HoodieStorageLayout.java) storage formats/layouts. Tables can be bucketed, consistent hashed or organized by having data laid out in order of arrival (default).
+3. Hudi already allows plug-ability and customization at various layers like record merger, indexes and other core write/read paths.
+4. It's very standard practice in databases to allow multiple storage backends (MySQL supports myISAM, innodb/btree, myrocks/lsm). This may be crucial step towards the database northstar vision.
+5. As a long time member of the Hudi community and open-source enthusiast, I think supporting other table formats, even existing ones like Apache Iceberg or Delta Lake, benefits those communities as well.
+    1. For e.g. the Hudi Streamer tool being used at our data lake (and hundreds more) for ingestion/incremental ETL can also benefit other communities.
+    2. Hudi provides out-of-box automatic table management that is manual in projects like Iceberg. With such an implementation, common data lake services can be reused across formats.
+    3. Hudi's high performance writer path can be extended to other formats (to the extent possible, that is not dependent on features like indexing that is only in Hudi's native format)
+    4. there are more such services and functionalities to be unlocked.
+
+
+Some non-technical reasons:
+1. Though Hudi is clearly defined as a platform over the years, there is so much vendor attention in the space for the past couple years, where Hudi is minimized to a table format and compared. This change will help highlight the value of Hudi's open software services, beyond just open formats.
+2. It may be controversial to say this. But, the project has been facing a lot of vendor FUD due to different vendors supporting different table formats. It is neither in the interest nor the business of the project community to be part of vendor wars. Opening up the table format layer to different implementations avoids these distractions for regular OSS contributors with no vendor interests, and helps focus on open-source software design and development.
+
+  
+## **Implementation**
+
+The main implementation step here is to create abstraction called TableFormatPlugin which handles table format operations such as 
```

Comment:
Are we gonna provide a set of APIs for plugbility or we have some impls like for Iceberg.

If we deem Hudi as a database, pluggable table format makes sense for me, my concern is the limited energy the community have. The current native Hudi table format already got a lot of complexity and maintainance is already hard, do we have additional energy to contribute to other table formats now?

Reply:
I am not proposing contributing to any projects. In my use case, I want to implement timeline and metadata table backed by nosql like dynamodb and not s3, for speed. This rfc just makes those diff impls possible

Reply:
> I want to implement timeline and metadata table backed by nosql like dynamodb and not s3

Something like the enforce of existing metaserver? Maybe we rename the title of the RFC, the current one is misleading.

Reply:
The point @danny0405 makes has some merit. I don't know if the Hudi community needs to maintain any of these implementations. I think it adds unnecessary dev and testing overhead. Can we limit the scope to just keeping the interfaces in the project. for e.g plenty of users have written custom payload mergers, not everything needs to be checked in and maintained within this repo

Reply:
Ok second thought. You are probably right. I was thinking in the lens of the dynamoDB timeline, which could may be live in `hudi-aws`. But, I agree that other table format plugins, can cause overhead for this codebase. I am happy to change the proposal to just have interfaces. I will keep our DyanmoDB timeline impl private, unless someone in the community wants it open. We can then discuss that separately.

Reply:
There are a bunch of hudi-extensions that are maintained like this in XTable.  
https://github.com/apache/incubator-xtable/tree/main/xtable-hudi-support/xtable-hudi-support-extensions/src/main/java/org/apache/xtable/hudi/extensions 

May be some of these can live there? We also maintain similar bridges for other table formats. We’d be happy to take a PR there.

Reply:
I also think the the implementation should NOT live within the Hudi project for the same reasons.

Reply:
Agree. @vinishjail97 / @yihua : Thanks a lot for the suggestion. This makes sense. Made the changes to the proposal to only keep the interface in Hudi while keeping the implementation for external formats in xtable.